source/common/ssl/connection_impl.cc

```diff
@@ -211,6 +211,7 @@ Network::ConnectionImpl::IoResult ConnectionImpl::doWriteToSocket() {
     // at the beginning of the buffer.
     if (inner_bytes_written > 0) {
       write_buffer_->drain(inner_bytes_written);
+      checkForLowWatermark();
```

Comment:
doWriteToSocket is also called in a couple of places so it turns out not to help.   I think it may be more likely that there will be more calls to doWriteToSocket (which would have to know to drain) than buffer-draining outside that function so I'm inclined to leave it as is.